Windows Compatibility?

KAllan357 opened the pull request:
What needs to change to make this provider work on Windows?

Some thoughts:
- current symlink / symlink_it_up
- Unzip commands

KAllan357 commented:
First issue popped up - `::File.readlink` is not implemented on Windows compiled Rubies.

At the moment, thinking about adding the following:

```
def readlink(path)
  if node[:platform] =~ /windows/
    require 'chef/win32/file'
    Chef::ReservedNames::Win32::File.readlink(path)
  end
  ::File.readlink(path)
end

def symlink?(path)
  if node[:platform] =~ /windows/
    require 'chef/win32/file'
    Chef::ReservedNames::Win32::File.symlink?(path)
  end
  ::File.symlink?(path)
end
```

KAllan357 committed:
update fixutres and add a new windows fixture

KAllan357 committed:
add a new readlink method that uses a Chef class to read windows symlinks

KAllan357 committed:
use file.join to avoid mishaps

KAllan357 committed:
formatting

ivey commented:
This looks good. Does Chef::ReservedNames::Win32::File.symlink? work on all Windows or do we need to document certain versions?

Is unzip not an issue?

KAllan357 committed:
use the windows cookbook to unzip files when windows

KAllan357 committed:
verify windows > vista

KAllan357 committed:
add docs

KAllan357 committed:
more doc edits

KAllan357 commented:
@ivey Please take a look. I added another check in the `load_current_resource` method that catches the raised exception thrown by Chef when the version of Windows you are using doesn't support symlinks. I then manually abort the Chef run. 

I think this makes the most sense -> this resource is highly dependent on symlinks and its probably better to just have a hard fail instead of a decent amount of effort for finding a workaround on Windows.

I also added the windows cookbook as a dependency and use the `windows_zipfile` resource for the second case in `extract_artifact!`. So far, I've seen that the Chef installation includes and adds a `tar.exe` to the Windows PATH so the first case should be handled.

ivey commented:
Looks good to me

reset commented:
[The Symlink file functions work on Vista and higher](https://github.com/opscode/chef/blob/master/lib/chef/win32/file.rb#L67-L68)

Let's make sure the metadata of the cookbook reflects this requirement

reset commented on providers/deploy.rb:
```diff
@@ -45,6 +45,15 @@ def load_current_resource
     Chef::Application.fatal! "The name attribute for this resource is significant, and there cannot be whitespace. The preferred usage is to use the name of the artifact."
   end
 
+  if windows?
```
this bit shouldn't be necessary if we lock our supported versions in the metadata

reset commented on libraries/chef_artifact.rb:
```diff
@@ -61,17 +61,35 @@ def get_actual_version(node, artifact_location, ssl_verify=true)
+        # @param  path [String] the path to a symlink
+        # 
+        # @return [String] the path that the symlink points to
+        def readlink(platform, path)
```
this smells like a function that could be extracted out into a mixin if we use it in other places

reset commented on metadata.rb:
```diff
@@ -5,7 +5,8 @@
 version          "1.3.1"
 
-%w{ centos redhat fedora }.each do |os|
+%w{ centos redhat fedora windows }.each do |os|
```
we need to hardlock on a windows version since we can't support ALL windows versions (because of symlinks)

reset commented on providers/deploy.rb:
```diff
@@ -557,4 +572,9 @@ def write_manifest
+  end
+
+  # @return [Fixnum or nil]
+  def windows?
```
another useful function that might fit nice in a mixin at a later date

reset commented:
@KAllan357 looks good man!

KAllan357 committed:
use metadatas supports feature to make sure we're using Vista or greater

KAllan357 committed:
add ruby docs to the retrieve_from_nexus library method

KAllan357 committed:
move some methods to a new mixin. some more minor changes for windows.

KAllan357 commented:
@ivey and @reset, please take a look at 5b600ae1e7208d8091ec967c87ae2b992630917b. I took the mixin suggestion and made a few more changes to get Windows support working.

KAllan357 committed:
add a comment to remember about CHEF-3960

KAllan357 committed:
conditionals are hard

KAllan357 committed:
remove platform attribute, use Chef::Platform.windows? method

ivey commented on fixtures/artifact_test/recipes/default.rb:
```diff
@@ -8,15 +8,13 @@
-user "artifact" do
-  group "artifact"
-end
+user "artifacts"
```
should this be `"artifact"`

ivey commented:
:+1:
